examples/topic_mirror.rs

```diff
@@ -53,34 +52,61 @@ fn create_consumer(brokers: &str, group_id: &str, topic: &str) -> LoggingConsume
         .create_with_context::<LoggingContext, LoggingConsumer>(LoggingContext)
         .expect("Consumer creation failed");
 
-    consumer.subscribe(&[topic]).expect("Can't subscribe to specified topic");
+    consumer
+        .subscribe(&[topic])
+        .expect("Can't subscribe to specified topic");
 
     consumer
 }
 
 struct MirrorProducerContext {
-    source_consumer: Arc<LoggingConsumer>
+    source_consumer: Arc<LoggingConsumer>,
 }
 
 impl ClientContext for MirrorProducerContext {}
 impl ProducerContext for MirrorProducerContext {
-    type DeliveryOpaque = TopicPartitionList;
+    type DeliveryOpaque = ();
 
-    fn delivery(&self, delivery_result: &Result<BorrowedMessage, (KafkaError, BorrowedMessage)>, delivery_opaque: TopicPartitionList) {
+    fn delivery(&self, delivery_result: &Result<BorrowedMessage, (KafkaError, BorrowedMessage)>, _delivery_opaque: ()) {
         match delivery_result {
             Ok(m) => {
-                info!("Delivered copy of {:?} to {}", delivery_opaque, m.topic());
-                self.source_consumer.store_offset_list(&delivery_opaque);
-            },
-            Err((e, m)) => error!("Failed to deliver to {}: {:?}", m.topic(), e)
+                debug!("Delivered copy of {:?} to {}", m.payload().unwrap(), m.topic());
+                // Question(blt) -- This call to `store_offset` will eventually
+                // cause offsets to be committed out of order. That will cause
+                // data loss. Should, then, the author cobble together a MinHeap
+                // (or similar) approach to commit only when all the suitable
+                // offsets have come through -- implying storage here -- or rely
+                // on rdkafka to take care of this?
+                //
+                // Either approach, potentially, interacts with the retry
+                // question below.
+                while let Err(_) = self.source_consumer.store_offset(&m) {
```

Comment:
Here `m` is the message that was just sent to Kafka (the copy, not the original one), so you'd be committing the offset of the copy, which is not what you want. The `TopicPartitionList` stores the location (offset and partition) of the original message, which is the one you want to commit.

Reply:
Ah, d'oh. Of course.